src/CMakeLists.txt

```diff
@@ -255,18 +298,21 @@ add_executable(avrdude
     whereami.h
     "${EXTRA_WINDOWS_SOURCES}"
     )
-
-target_link_libraries(avrdude PUBLIC libavrdude)
+if (sharedlib IN_LIST BUILT_LIBS)
+    target_link_libraries(avrdude PUBLIC sharedlib)
+else()
+    target_link_libraries(avrdude PUBLIC staticlib)
+endif()
```

Comment:
Should we retain the old behaviour of linking the executable with the static library? Or should we make the executable dynamically-linked (like I've done here) ?

@dl8dtl @mariusgreuel

Reply:
May I ask, why would one _not_ have binaries with statically linked libraries? It makes it way easier to bundle the official Avrdude release binaries without having to think about having the right drivers installed.
A binary that "just works" is what I personally prefer.

Reply:
Because it entirely defeats the idea of a shared/dynamic library?

To me, it seems this is really a philosophical issue. On Unix, 30 years ago, everyone was asking for shared libraries, because they can be *shared* between applications. That's why they have the suffix `.so`, for a "shared object".

On Windows, everyone tends to ship their own version of the DLL anyway. That completely defeats the idea of sharing something. It's then really just a complicated way of doing what could have been achieved with a static linking in the first place.

For that, I'm perfectly fine if we separate it that way: link statically on Windows, and used shared libs an Unix.

Reply:
Surely, for AVRDUDE, it's not really a concern, unless we might have a second consumer of the shared lib some day. Even running multiple AVRDUDE processes doesn't gain from a shared library since the "text" (program code) will be shared between these processes anyway.

Reply:
> It makes it way easier

This is the whole point here. Static libraries are so much easier to use. And it is not just bundling packages. Shared objects or DLLs are a lot of work. You either have to keep your interfaces stable, or you better make sure you get your versioning straight. On Windows, this is why everybody uses COM, i.e. you have immutable interfaces that work very much like pure virtual C++ classes, including all the benefits this implies. And this is what I meant earlier, when I said the avrdude codebase is in no shape for a shared object: We currently have no abstraction at all, let alone immutable interfaces.

I think we should focus on adding features to avrdude, not maintaining shared objects that are not going to be used.

Reply:
> > It makes it way easier
> 
> This is the whole point here. Static libraries are so much easier to use.

I think we are in strong disagreement here. ;-)

> And it is not just bundling packages. Shared objects or DLLs are a lot of work.

I don't think they are more work than static ones. (OK, Windows DLLs seem to be more work, since it seems they cannot refer to global symbols from the resulting executable. That's another matter.)

> You either have to keep your interfaces stable,

We need  the same for a static library. If we flip the API all the time, nobody could rely on the static version as well. The only difference is, for a static library, the failure becomes already obvious at link time – but it remains a failure nevertheless.

If the API is only getting extended from release to release, there's no fear of incompatibilty and versioning.

Segregating the library backend has been a long-standing project, and I wouldn't want to give it up. It helps getting some structure into the code that hasn't been there before. Dynamic libraries are just a side-product then. If you don't like them, don't use them.

Reply:
> A binary that "just works" is what I personally prefer.

So do I. 

But IIRC, the executable that is built currently by using autotools is *not* fully static.

```
$ ldd `command -v avrdude`
	linux-vdso.so.1 (0x00007fffeadf4000)
	libusb-1.0.so.0 => /usr/lib/libusb-1.0.so.0 (0x00007f2ff5be5000)
	libusb-0.1.so.4 => /usr/lib64/libusb-0.1.so.4 (0x00007f2ff5bde000)
	libftdi1.so.2 => /usr/lib64/libftdi1.so.2 (0x00007f2ff5bcd000)
	libelf.so.1 => /usr/lib/libelf.so.1 (0x00007f2ff5bb2000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f2ff5b91000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007f2ff5a4c000)
	libhidapi-libusb.so.0 => /usr/lib64/libhidapi-libusb.so.0 (0x00007f2ff5a40000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f2ff587a000)
	libudev.so.1 => /usr/lib/libudev.so.1 (0x00007f2ff5852000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f2ff5c97000)
	libz.so.1 => /usr/lib/libz.so.1 (0x00007f2ff5838000)
	librt.so.1 => /usr/lib/librt.so.1 (0x00007f2ff582d000)
```

See? It still depends on `lib{usb,ftdi1,elf,hidapi}`

If we decide to make `avrdude` static, then we'll have to link those (and the libc too, IMO) statically.

---

But that's not what my original question is about my original question is about dynamically linking to `libavrdude`.

My point was that, since we're dynamically linking to the other libraries, why not link dynamically to `libavrdude`?

That was the rationale behind this code -
```cmake
if (sharedlib IN_LIST BUILT_LIBS)
    target_link_libraries(avrdude PUBLIC sharedlib)
else()
    target_link_libraries(avrdude PUBLIC staticlib)
endif()
```

and it caused the resulting executable to be dynamically linked with `libavrdude`
```
$ ldd `command -v avrdude`
	linux-vdso.so.1 (0x00007ffe321f6000)
	libavrdude.so.1 => /usr/lib64/libavrdude.so.1 (0x00007fcadf372000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007fcadf1ac000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007fcadf067000)
	libelf.so.1 => /usr/lib/libelf.so.1 (0x00007fcadf04c000)
	libusb-0.1.so.4 => /usr/lib64/libusb-0.1.so.4 (0x00007fcadf045000)
	libusb-1.0.so.0 => /usr/lib/libusb-1.0.so.0 (0x00007fcadf026000)
	libhidapi-libusb.so.0 => /usr/lib64/libhidapi-libusb.so.0 (0x00007fcadf01a000)
	libftdi1.so.2 => /usr/lib64/libftdi1.so.2 (0x00007fcadf009000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007fcadf425000)
	libz.so.1 => /usr/lib/libz.so.1 (0x00007fcadefef000)
	libudev.so.1 => /usr/lib/libudev.so.1 (0x00007fcadefc7000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007fcadefa6000)
	librt.so.1 => /usr/lib/librt.so.1 (0x00007fcadef9b000)
```

Reply:
> the executable that is built currently by using autotools is not fully static

Lets talk about CMake, but even with the Automake, this is per design. Jörg and I actually talked briefly about this and thought this would make the most sense. Reason being:

- On a Un*x system, we have a controlled environment, i.e. a package manager with dependency management. avrdude will be build for a specific environment, i.e. we know which shared object are present and avrdude will work with. Alternatively, it is relatively easy to build locally, where again avrdude is run in a matching environment. That is also the reason that the release artifacts do not contain any Un*x binaries, because dynamically linked executables won't just work in an arbitrary Un*x system.
- On Windows, we do not have a controlled environment. avrdude is typically distributed via a download, so we do not know the version of Windows it is run, or which or which versions of the DLLs are present. Hence, we link everything (i.e. libusb, libhid, libftdi, libelf) but the system DLLs static. The crt libraries are static as well because we would have to distribute them otherwise.

Now, with the Arduino team, we have a new request here: We want to be able to install avrdude in an environment we do not control: For instance, download the Arduino IDE and execute it on any Linux system. So now, we have the same set of problems as on Windows, and now Linux people do something similar what Windows people are used to: 1) Either make sure the dependencies are met through some sort of installer, or 2) install shared objects side-by-side with the application, or 3) link everything (but the system/crt libraries) statically.

For a standalone application like avrdude, 3) is a good choice.

Reply:
> My point was that, since we're dynamically linking to the other libraries, why not link dynamically to libavrdude?

Well, if you specify `BUILD_SHARED_LIBS=1` you should.